### PR TITLE
LIF refractory floating point

### PR DESCRIPTION
**Pull request by ssgier**
Issue Number:  #282

Objective of pull request: Add LIF process with refractory period

## Pull request checklist

Your PR fulfills the following requirements:
- [x] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [x] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

- No LIF process with refractory period is available

## What is the new behavior?

- LIF process with refractory period is available. CPU/Floating point process model that implements the process is available.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Supplemental information

As discussed with @mathisrichter, I created a new process based on the LIF process with reset. This change currently only includes the floating point process model in Python as a start. It could serve as a starting point to discuss/implement other process models.

### CODE REVIEW

**Comment by mathisrichter** [comment]
Oh boy, sorry @ssgier for the long delay. I have not been monitoring Github for a while and we overlooked this PR.
I have now assigned a few people from my team to review it.

Hopefully, we can review and get this merged soon now.
My apologies!

File: src/lava/proc/lif/process.py [hunk 1]
```diff
@@ -347,3 +347,75 @@ def __init__(
...
+            du=du,
+            dv=dv,
+            bias_mant=bias_mant,
+            bias_exp=bias_exp,
+            vth=vth,
+            name=name,
+            log_config=log_config,
+        )
+
+        if refractory_period < 1:
```
Comment (weidel-p) [inline src/lava/proc/lif/process.py:417]: In my understanding, a refractory period of 0 should be valid and mean "no refractory period".

**Review by weidel-p — CHANGES_REQUESTED** [review]
Thanks for this clean implementation there is not much to ask for.  

The only thing is, that I would like to change the default value to 0 instead of 1. In your current implementation, 1 means "no refractoryness" and yields the same results as the normal LIF. In my understanding, 1 should mean that the voltage is clamped to zero for one timestep after spiking. Hence I would change the line

```non_refractory = self.refractory_period_end <= self.time_step```

to 

``` non_refractory = self.refractory_period_end < self.time_step ``` 

and allow refractory_period to be 0 but not negative.

File: src/lava/proc/lif/process.py (hunk 1 above)
Comment (ssgier) [reply src/lava/proc/lif/process.py:417]: The reason why I wrote the check like this is that, as you pointed out, a refractory period of zero would result in the same behavior as the normal LIF, which might be unintended by the user (why would he not use the normal LIF instead?). Do you want me to adjust this nonetheless?

Comment (weidel-p) [reply src/lava/proc/lif/process.py:414]: no, I think that's fine. Thanks.

File: tests/lava/proc/lif/test_models.py [hunk 2]
```diff
@@ -824,3 +825,37 @@ def test_fixed_model(self):
...
+        v_logger = io.sink.Read(buffer=num_steps)
+        v_logger.connect_var(lif_refractory.v)
+
+        lif_refractory.run(condition=RunSteps(num_steps),
+                           run_cfg=Loihi2SimCfg(select_tag="floating_pt"))
+
+        v = v_logger.data.get()
+        lif_refractory.stop()
+
+        # voltage is expected to remain at reset level for two time steps
```
Comment (PhilippPlank) [inline tests/lava/proc/lif/test_models.py:857]: Please start comments with capital letters to stay consistent.

**Review by weidel-p — APPROVED** [review]

**Pull request merged**